<a href="https://colab.research.google.com/github/loresiensis/data-analysis-and-nlp/blob/main/Traduccion_automatica_y_evaluacion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Texto original

In [ ]:
original = 'The obvious answer is that since we can, of course, make decisions for ourselves, there is no reason why these shouldn’t include ourselves in the future. Indeed, we make such decisions all the time, when we sign up for twenty-five year mortgages, pension plans, to have and to hold till death us do part, or even just for a two-year work contract. Alongside this duty to make good our promises, though, there must also surely be a concomitant right to change our minds as our circumstances and beliefs change. Many people, for example, say things that begin ‘shoot me if I ever …’, especially when young. And although it is often just a figure of speech, it is frequently said with the utmost sincerity, and often by people who are of the age of majority and so considered adults capable of making decisions about their own futures. To hold people to these vows, however, would be ridiculous. However, the fact that we do think it reasonable to change our minds should make us see these other long term commitments as also being less than absolute. For the plain truth is that we change. In a very real sense, we are not the same people that we were many years in the past. So when we make promises on behalf of our future selves, we are to some degree making promises for someone other than who we now are. And that means our promises should not be seen as morally binding.'

In [ ]:
original_phrases = [x.strip() for x in original.split('.')[:-1]]
original_phrases

['The obvious answer is that since we can, of course, make decisions for ourselves, there is no reason why these shouldn’t include ourselves in the future',
 'Indeed, we make such decisions all the time, when we sign up for twenty-five year mortgages, pension plans, to have and to hold till death us do part, or even just for a two-year work contract',
 'Alongside this duty to make good our promises, though, there must also surely be a concomitant right to change our minds as our circumstances and beliefs change',
 'Many people, for example, say things that begin ‘shoot me if I ever …’, especially when young',
 'And although it is often just a figure of speech, it is frequently said with the utmost sincerity, and often by people who are of the age of majority and so considered adults capable of making decisions about their own futures',
 'To hold people to these vows, however, would be ridiculous',
 'However, the fact that we do think it reasonable to change our minds should make us see

In [ ]:
!pip install transformers[sentencepiece] -Uq

In [ ]:
from transformers import pipeline
model_checkpoint = "Helsinki-NLP/opus-mt-tc-big-en-es"
translation_model = pipeline("translation", model=model_checkpoint)

/usr/local/lib/python3.8/dist-packages/transformers/models/marian/tokenization_marian.py:194: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")


Traducción con el modelo Helsinki-NLP

In [ ]:
translation_model(original_phrases)

[{'translation_text': 'La respuesta obvia es que, dado que podemos, por supuesto, tomar decisiones por nosotros mismos, no hay razón por la que estos no deban incluirnos en el futuro.'},
 {'translation_text': 'De hecho, tomamos tales decisiones todo el tiempo, cuando nos inscribimos en hipotecas de veinticinco años, planes de pensiones, para tener y mantener hasta la muerte nos hacen parte, o incluso sólo por un contrato de trabajo de dos años.'},
 {'translation_text': 'Sin embargo, junto con este deber de cumplir nuestras promesas, también debe haber un derecho concomitante de cambiar nuestras mentes a medida que cambian nuestras circunstancias y creencias.'},
 {'translation_text': 'Muchas personas, por ejemplo, dicen cosas que comienzan a “dispararme si alguna vez...”, especialmente cuando son jóvenes.'},
 {'translation_text': 'Y aunque a menudo es sólo una figura retórica, a menudo se dice con la mayor sinceridad, y a menudo por personas que son de la mayoría de edad y por lo tanto 

In [ ]:
traducciones1 = translation_model(original_phrases)
traducciones1 = [x['translation_text'] for x in traducciones1]
traducciones1

['La respuesta obvia es que, dado que podemos, por supuesto, tomar decisiones por nosotros mismos, no hay razón por la que estos no deban incluirnos en el futuro.',
 'De hecho, tomamos tales decisiones todo el tiempo, cuando nos inscribimos en hipotecas de veinticinco años, planes de pensiones, para tener y mantener hasta la muerte nos hacen parte, o incluso sólo por un contrato de trabajo de dos años.',
 'Sin embargo, junto con este deber de cumplir nuestras promesas, también debe haber un derecho concomitante de cambiar nuestras mentes a medida que cambian nuestras circunstancias y creencias.',
 'Muchas personas, por ejemplo, dicen cosas que comienzan a “dispararme si alguna vez...”, especialmente cuando son jóvenes.',
 'Y aunque a menudo es sólo una figura retórica, a menudo se dice con la mayor sinceridad, y a menudo por personas que son de la mayoría de edad y por lo tanto considerados adultos capaces de tomar decisiones sobre su propio futuro.',
 'Sostener a la gente a estos voto

Traducción con [DeepL](https://www.deepl.com/translator)

In [ ]:
traducciones2 = ['La respuesta obvia es que, puesto que podemos, por supuesto, tomar decisiones por nosotros mismos, no hay razón para que éstas no nos incluyan en el futuro',
                 'De hecho, tomamos tales decisiones todo el tiempo, cuando firmamos hipotecas a veinticinco años, planes de pensiones, para tener y mantener hasta que la muerte nos separe, o incluso simplemente para un contrato de trabajo de dos años',
                 'Sin embargo, junto a este deber de cumplir nuestras promesas, debe existir también el derecho concomitante a cambiar de opinión a medida que cambian nuestras circunstancias y creencias',
                 'Muchas personas, por ejemplo, dicen cosas que empiezan por "dispárame si alguna vez...", especialmente cuando son jóvenes',
                 'Y aunque a menudo es sólo una forma de hablar, con frecuencia se dice con la mayor sinceridad, y a menudo por personas que han alcanzado la mayoría de edad y, por tanto, se consideran adultos capaces de tomar decisiones sobre su propio futuro',
                 'Sin embargo, sería ridículo obligar a la gente a hacer estos votos',
                 'Sin embargo, el hecho de que consideremos razonable cambiar de opinión debería hacernos ver que estos otros compromisos a largo plazo también son menos que absolutos',
                 'La verdad es que cambiamos',
                 'En un sentido muy real, no somos las mismas personas que hace muchos años',
                 'Por eso, cuando hacemos promesas en nombre de nuestro yo futuro, en cierta medida estamos haciendo promesas para alguien distinto de lo que somos ahora',
                 'Y eso significa que nuestras promesas no deben considerarse moralmente vinculantes.']

Evaluación

In [ ]:
reference_translation = ['La respuesta obvia es que, dado que podemos, por supuesto, tomar decisiones por nosotros mismos, no hay razón para que éstas no nos incluyan en el futuro',
                         'De hecho, tomamos tales decisiones todo el tiempo, cuando firmamos hipotecas a veinticinco años, planes de pensiones, uniones hasta que la muerte nos separe, o incluso simplemente un contrato de trabajo de dos años',
                         'Sin embargo, junto a este deber de cumplir nuestras promesas, debe existir también el derecho concomitante a cambiar de opinión a medida que cambian nuestras circunstancias y creencias',
                         'Muchas personas, por ejemplo, dicen cosas que empiezan por "dispárame si alguna vez...", especialmente cuando son jóvenes',
                         'Y aunque a menudo es sólo una forma de hablar, con frecuencia se dice con la mayor sinceridad, y a menudo por personas que han alcanzado la mayoría de edad y se consideran, por tanto, adultos capaces de tomar decisiones sobre su propio futuro',
                         'A pesar de esto, sería ridículo cumplir con estas peticiones',
                         'No obstante, el hecho de que consideremos razonable cambiar de opinión debería hacernos ver que estos otros compromisos a largo plazo tampoco son absolutos',
                         'Lo cierto es que cambiamos',
                         'En un sentido muy real, no somos las mismas personas que hace muchos años',
                         'Por eso, cuando hacemos promesas en nombre de nuestro yo futuro, en cierta medida estamos haciendo promesas para alguien distinto de lo que somos ahora',
                         'Y eso significa que nuestras promesas no deberían considerarse moralmente obligatorias.']

In [ ]:
!pip install evaluate -Uq

In [ ]:
!pip install jiwer bert_score git+https://github.com/google-research/bleurt.git -Uq

  Preparing metadata (setup.py) ... done


WER

In [ ]:
from evaluate import load
wer = load("wer")

Score de Helsinki-NLP

In [ ]:
wer_score1 = wer.compute(predictions=traducciones1, references=reference_translation)
wer_score1

0.3723849372384937

Score de DeepL

In [ ]:
wer_score2 = wer.compute(predictions=traducciones2, references=reference_translation)
wer_score2

0.13389121338912133

WER es una métrica de error, con lo cual DeepL sería mejor que Helsinki-NLP.

BLEU

In [ ]:
bleu = load("bleu")

Score de Helsinki-NLP

In [ ]:
bleu_score1 = bleu.compute(predictions=traducciones1, references=reference_translation)
bleu_score1

{'bleu': 0.5559318449726542,
 'precisions': [0.7235494880546075,
  0.5921985815602837,
  0.5129151291512916,
  0.4346153846153846],
 'brevity_penalty': 1.0,
 'length_ratio': 1.0932835820895523,
 'translation_length': 293,
 'reference_length': 268}

Score de DeepL

In [ ]:
bleu_score2 = bleu.compute(predictions=traducciones2, references=reference_translation)
bleu_score2

{'bleu': 0.8537871927818911,
 'precisions': [0.9094202898550725,
  0.8679245283018868,
  0.8346456692913385,
  0.8065843621399177],
 'brevity_penalty': 1.0,
 'length_ratio': 1.0298507462686568,
 'translation_length': 276,
 'reference_length': 268}

En la métrica BLEU cuanto más alto el valor se considera una mejor traducción, así que nuevamente gana la de DeepL.

METEOR

In [ ]:
meteor = load("meteor")

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


Score de Helsinki-NLP

In [ ]:
meteor_score1 = meteor.compute(predictions=traducciones1, references=reference_translation)
meteor_score1

{'meteor': 0.7384653422941533}

Score de DeepL

In [ ]:
meteor_score2 = meteor.compute(predictions=traducciones2, references=reference_translation)
meteor_score2

{'meteor': 0.8726770078352043}

Nuevamente DeepL tiene un mejor score.

BERT Score

In [ ]:
bertscore = load("bertscore")

Score de Helsinki-NLP

In [ ]:
bert_score1 = bertscore.compute(predictions=traducciones1, references=reference_translation,lang="es")
bert_score1

{'precision': [0.9262920022010803,
  0.8642554879188538,
  0.9407663345336914,
  0.8883928060531616,
  0.9235420227050781,
  0.7303110361099243,
  0.9181063771247864,
  0.7879133820533752,
  0.8841434121131897,
  0.915709376335144,
  0.9280824661254883],
 'recall': [0.934088408946991,
  0.8965352773666382,
  0.9444389343261719,
  0.8823687434196472,
  0.9132331609725952,
  0.7506001591682434,
  0.9445184469223022,
  0.8393476009368896,
  0.9478656649589539,
  0.9194135665893555,
  0.9250270128250122],
 'f1': [0.9301738739013672,
  0.8800994753837585,
  0.9425990581512451,
  0.8853706121444702,
  0.9183586239814758,
  0.7403165698051453,
  0.9311251640319824,
  0.8128176331520081,
  0.9148963093757629,
  0.9175577163696289,
  0.9265522360801697],
 'hashcode': 'bert-base-multilingual-cased_L9_no-idf_version=0.3.12(hug_trans=4.26.1)'}

Score de DeepL

In [ ]:
bert_score2 = bertscore.compute(predictions=traducciones2, references=reference_translation,lang="es")
bert_score2

{'precision': [0.996508777141571,
  0.9485462307929993,
  1.0,
  1.0,
  0.9908040761947632,
  0.7927690148353577,
  0.9618679285049438,
  0.8840469717979431,
  1.0,
  0.9999999403953552,
  0.954016387462616],
 'recall': [0.996508777141571,
  0.9704412817955017,
  1.0,
  1.0,
  0.9908041954040527,
  0.8236273527145386,
  0.9801724553108215,
  0.8840469717979431,
  1.0,
  0.9999999403953552,
  0.943108856678009],
 'f1': [0.996508777141571,
  0.959368884563446,
  1.0,
  1.0,
  0.990804135799408,
  0.8079035878181458,
  0.9709339141845703,
  0.8840469717979431,
  1.0,
  0.9999999403953552,
  0.9485312700271606],
 'hashcode': 'bert-base-multilingual-cased_L9_no-idf_version=0.3.12(hug_trans=4.26.1)'}

Para tener un poco más claras las métricas anteriores sacamos una media y desviación típica

In [ ]:
import numpy as np
np.mean(bert_score1['f1']),np.std(bert_score1['f1'])

(0.8908970247615468, 0.0588250558142322)

In [ ]:
np.mean(bert_score2['f1']),np.std(bert_score2['f1'])

(0.9598270437934182, 0.05864772600020361)

La desviación típica es muy similar en ambos, pero la media de F1 es más alta para la traducción de DeepL

BLEURT

In [ ]:
bleurtscore = load("bleurt")

Score de Helsinki-NLP

In [ ]:
bleurt_score1 = bleurtscore.compute(predictions=traducciones1, references=reference_translation)
bleurt_score1

{'scores': [0.5979195833206177,
  0.2204456776380539,
  0.42691218852996826,
  0.529697597026825,
  0.18949896097183228,
  -0.11392396688461304,
  0.30275481939315796,
  0.4065021276473999,
  0.6438672542572021,
  0.27149301767349243,
  0.3750607371330261]}

Score de DeepL

In [ ]:
bleurt_score2 = bleurtscore.compute(predictions=traducciones2, references=reference_translation)
bleurt_score2

{'scores': [0.9293946623802185,
  0.4424673318862915,
  0.9909730553627014,
  0.9902089238166809,
  0.8837072253227234,
  -0.08438979089260101,
  0.3173651695251465,
  0.41629558801651,
  1.0253393650054932,
  0.9967148900032043,
  0.4884430766105652]}

De nuevo, nos conviene sacar una media y desviación típica de estos scores

In [ ]:
np.mean(bleurt_score1['scores']),np.std(bleurt_score1['scores'])

(0.35002072697336023, 0.20360727051565808)

In [ ]:
np.mean(bleurt_score2['scores']),np.std(bleurt_score2['scores'])

(0.672410863366994, 0.35598950640714944)

La desviación típica de la traducción de DeepL es algo mayor que la del modelo Helsinki-NLP, pero la media también es considerablemente mejor.

Por último podemos ver todas las métricas en una tabla:

||Helsinki-NLP/opus-mt-tc-big-en-es|DeepL|
|---|---|---|
| WER | 0.37 | 0.13|
| BLEU | 0.55 | 0.85|
| METEOR |0.73 | 0.87|
| BERT Score | 0.89 (0.05) | 0.95 (0.05)|
| BLEURT | 0.35 (0.20) | 0.67 (0.35)|


